backend-api/src/functions/adapters/dynamoDbAdapter.ts

```diff
@@ -0,0 +1,118 @@
+import {
+  ConditionalCheckFailedException,
+  DynamoDBClient,
+  PutItemCommand,
+  PutItemCommandInput,
+  QueryCommand,
+  QueryCommandInput,
+  QueryCommandOutput,
+} from "@aws-sdk/client-dynamodb";
+import { CreateSessionAttributes } from "../services/sessionService/sessionService";
+import { NodeHttpHandler } from "@smithy/node-http-handler";
+import { NativeAttributeValue, unmarshall } from "@aws-sdk/util-dynamodb";
+
+const sessionStates = {
+  ASYNC_AUTH_SESSION_CREATED: "ASYNC_AUTH_SESSION_CREATED",
+};
+
+export type DatabaseRecord = Record<string, NativeAttributeValue>;
+
+export class DynamoDbAdapter {
+  private readonly tableName: string;
+  private readonly dynamoDbClient = new DynamoDBClient({
+    region: process.env.REGION,
+    maxAttempts: 2,
+    requestHandler: new NodeHttpHandler({
+      connectionTimeout: 29000,
+      requestTimeout: 29000,
+    }),
+  });
+
+  constructor(tableName: string) {
+    this.tableName = tableName;
+  }
+
+  async getActiveSession(
+    subjectIdentifier: string,
+    attributesToGet: string[],
+  ): Promise<DatabaseRecord | null> {
+    const input: QueryCommandInput = {
+      TableName: this.tableName,
+      IndexName: "subjectIdentifier-timeToLive-index",
+      KeyConditionExpression:
+        "subjectIdentifier = :subjectIdentifier and :currentTimeInSeconds < timeToLive",
+      FilterExpression: "sessionState = :sessionState",
+      ExpressionAttributeValues: {
+        ":subjectIdentifier": { S: subjectIdentifier },
+        ":sessionState": { S: sessionStates.ASYNC_AUTH_SESSION_CREATED },
+        ":currentTimeInSeconds": {
+          N: this.getTimeNowInSeconds().toString(),
+        },
+      },
+      ProjectionExpression: this.formatAttributesToGet(attributesToGet),
+      Limit: 1,
+      ScanIndexForward: false,
+    };
+
+    const queryCommandOutput: QueryCommandOutput =
+      await this.dynamoDbClient.send(new QueryCommand(input));
+
+    const items = queryCommandOutput.Items;
+    if (!items || items.length === 0) {
+      return null;
+    }
+
+    return unmarshall(items[0]);
```

Comment:
This is cool I had not heard of this before!

Reply:
I want to use `marshall()` when we create the session but that will be done in a separate PR as it requires a bit of testing to ensure all is the same